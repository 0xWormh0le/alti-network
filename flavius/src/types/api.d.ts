type QueryParamValue = string | number | string[] | number[] | null | undefined | boolean
type QueryParams = { [k: string]: QueryParamValue }

type ApiClientReturn<TState extends any, TCallerArgs extends any> = [
  Maybe<TState>,
  Maybe<{ error: string }>,
  boolean,
  ApiCaller<TCallerArgs>
]

type ApiErrorCallback = (err: any) => void
type ApiSuccessCallback = (response: any) => void

interface ApiClientInit {
  defaultPageSize?: number
  pageCountCacheTtl?: number
  handleError?: ApiErrorCallback
  handleSuccess?: ApiSuccessCallback
  immediateTrigger?: boolean
}

interface ApiCaller<T> {
  call: (args?: T) => void
}

interface PaginationArgs {
  pageSize: number
  pageCountCacheTtl: number
  pageCountLastUpdated: number
}

interface MetaArgs<T, TParams = unknown> {
  handleError?: (err: Error) => void
  handleSuccess?: (rs: T) => void
  params?: (QueryParams & TParams) | undefined
  preventFirstCall?: boolean
  triggeringKeys?: string[]
}

interface InternalMetaArgs<T, TParams = unknown> extends MetaArgs<T, TParams> {
  trigger: boolean
  resetTrigger: () => void
  handleError: (err: Error) => void
  handleSuccess: (rs: T) => void
  params: (QueryParams & TParams) | undefined
}
interface ApiCallArgs<T, TArgs> {
  requestDetails?: TArgs
  meta?: MetaArgs<T>
}

type WithPagination<T> = T & {
  pageNumber: number
  pageSize: number
}
