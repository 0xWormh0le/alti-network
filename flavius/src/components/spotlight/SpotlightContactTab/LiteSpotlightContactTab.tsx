import React from 'react'
import cx from 'classnames'
import PersonAvatar from 'components/elements/PersonAvatar'
import Typography, { TypographyVariant } from 'components/elements/Typography'
import { truncateText } from 'util/helpers'
import CopyEmail from '../CopyEmail'
import '../SpotlightContactTab/SpotlightContactTab.scss'

interface LiteSpotlightContactTabProps {
  emailInfo?: {
    address: string
  }
  person: Maybe<LitePerson>
  isSelected: boolean
}

const LiteSpotlightContactTab: React.FC<LiteSpotlightContactTabProps> = (props) => {
  const { isSelected, person, emailInfo } = props
  return (
    <div
      className={cx('SpotlightContactTabItem', {
        'SpotlightContactTabItem--selected': isSelected
      })}>
      {person && (
        <PersonAvatar
          style={{ fontSize: '0.75em', marginRight: '1em' }}
          person={person}
          className='SpotlightContactTabItem__avatar'
          clickable={false}
        />
      )}
      {isSelected ? (
        <CopyEmail email={emailInfo?.address ?? ''}>
          <Typography variant={TypographyVariant.H4} component='div' className='SpotlightContactTabItem__email-text'>
            {truncateText(emailInfo?.address ?? '', 18)}
          </Typography>
        </CopyEmail>
      ) : (
        <Typography variant={TypographyVariant.H4} component='div' className='SpotlightContactTabItem__email-text'>
          {truncateText(emailInfo?.address ?? '', 18)}
        </Typography>
      )}
    </div>
  )
}

export default LiteSpotlightContactTab
