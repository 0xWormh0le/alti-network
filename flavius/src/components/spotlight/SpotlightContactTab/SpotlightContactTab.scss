@import 'styles/_shared.scss';

.SpotlightContactTabItem {
  display: flex;
  align-items: center;
  justify-content: center;

  &__email-text {
    line-height: 1;
  }

  &--selected {
    color: #222c3c;
  }

  &__external-marker {
    margin-left: 0.5em;
  }
}
