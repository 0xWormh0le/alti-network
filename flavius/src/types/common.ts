export enum UsageKind {
  personal = 'personal',
  work = 'work'
}

export enum UserKind {
  user = 'user',
  person = 'person',
  serviceAccount = 'serviceAccount'
}

export enum AccessLevel {
  member = 'member',
  manager = 'manager',
  owner = 'owner',
  admin = 'admin'
}

export enum RiskCategoryType {
  SHARING_RISKS = 'Sharings',
  RELATIONSHIP_RISKS = 'Relationships',
  ACTIVITY_RISKS = 'Actions',
  INFORMATIONAL_RISKS = 'Informatives'
}

export enum FileActivityType {
  ACCESS_GRANTED = 'ACCESS_GRANTED',
  CHANGE_FOLDER_PERMISSION = 'CHANGE_FOLDER_PERMISSION',
  CONTENT_ACCESS = 'CONTENT_ACCESS',
  DOWNLOAD = 'DOWNLOAD',
  FILE_WATERMARKED_DOWNLOAD = 'FILE_WATERMARKED_DOWNLOAD',
  ITEM_DOWNLOAD = 'ITEM_DOWNLOAD',
  ITEM_PREVIEW = 'ITEM_PREVIEW',
  ITEM_SHARED = 'ITEM_SHARED',
  ITEM_SHARED_CREATE = 'ITEM_SHARED_CREATE',
  ITEM_SHARED_UPDATE = 'ITEM_SHARED_UPDATE',
  ITEM_SYNC = 'ITEM_SYNC',
  SHARE = 'SHARE'
}

export enum PermissionRole {
  own = 'own',
  read = 'read',
  write = 'write'
}

export enum PermissionShared {
  external = 'external',
  internal = 'internal',
  none = 'none'
}

export enum PermissionType {
  anyone = 'anyone',
  domain = 'domain',
  user = 'user'
}

export enum PermissionItemStatus {
  active = 'ACTIVE',
  pending = 'PENDING',
  removed = 'REMOVED',
  cannotBeRemoved = 'CANNOT_BE_REMOVED'
}

export enum SensitiveContentStatus {
  DETECTION_PENDING = 'pending',
  FOUND_SENSITIVE_CONTENT = 'yes',
  FOUND_NO_SENSITIVE_CONTENT = 'no',
  FAILED_DETECTION = 'failed'
}

export enum FileGridNavType {
  TIMELINE = 'timeline',
  INTERNAL = 'internal',
  EXTERNAL = 'external',
  SENSITIVE = 'sensitive',
  FILES_IN_FOLDER = 'filesInFolder',
  DETAILS = 'details'
}
