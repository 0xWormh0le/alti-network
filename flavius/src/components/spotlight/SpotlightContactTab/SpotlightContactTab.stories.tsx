import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import SpotlightContactTab, { SpotlightContactTabProps } from './SpotlightContactTab'

import { UsageKind, UserKind, AccessLevel } from 'types/common'

export default {
  title: 'Spotlight/SpotlightContactTab',
  component: SpotlightContactTab
} as Meta

const Template: Story<SpotlightContactTabProps> = (args) => <SpotlightContactTab {...args} />

export const Default = Template.bind({})
Default.args = {}

export const WithPerson = Template.bind({})
WithPerson.args = {
  person: {
    primaryEmail: {
      address: 'test@email.com',
      kind: UsageKind.personal,
      primary: true,
      accessCount: 0,
      riskCount: 0,
      lastDeletedPermissions: 0
    },
    recoveryEmail: {
      address: 'test@email.com',
      kind: UsageKind.personal,
      primary: true,
      accessCount: 0,
      riskCount: 0,
      lastDeletedPermissions: 0
    },
    name: {
      givenName: 'Test',
      familyName: 'User'
    },
    avatar: {
      url: '',
      url_etag: ''
    },
    accessCount: 0,
    riskCount: 0,
    emails: [
      {
        address: 'email2',
        accessCount: 1,
        kind: UsageKind.work,
        primary: true,
        riskCount: 1,
        lastDeletedPermissions: 0
      },
      {
        address: 'email3',
        primary: true,
        accessCount: 2,
        kind: UsageKind.personal,
        riskCount: 1,
        lastDeletedPermissions: 0
      }
    ],
    internalCount: 1,
    externalCount: 1,
    internal: false,
    altnetId: '',
    projectId: '',
    accessLevel: AccessLevel.member,
    userKind: UserKind.user,
    phones: [],
    externalIds: [],
    orgUnitPath: '',
    etag: '',
    isEnrolledInMFA: false,
    creationTime: 0,
    lastLoginTime: 0,
    lastModifiedTime: 0,
    notes: {
      content: '',
      content_type: ''
    }
  } as any
}
