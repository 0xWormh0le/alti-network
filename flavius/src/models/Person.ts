import { UserKind, AccessLevel } from 'types/common'
import UI_STRINGS from 'util/ui-strings'
import anonymousUserIcon from 'icons/anonymous-user.svg'
class Person implements IPerson {
  public accessLevel: AccessLevel
  public name: Name
  public projectId: string
  public primaryEmail: Email
  public userKind: UserKind
  public altnetId?: string
  public recoveryEmail?: Email
  public phones: Phone[]
  public externalIds: UserId[]
  public avatar?: Avatar
  public internal?: boolean // deprecated thanks to userKind
  public anonymous?: boolean
  public internalCount?: number
  public externalCount?: number
  public emails: Email[]
  public accessCount?: number
  public riskCount?: number
  public orgUnitPath?: string
  public isEnrolledInMFA?: boolean
  public etag?: string
  public creationTime?: number
  public lastLoginTime?: number
  public lastModifiedTime?: number
  public notes?: Notes
  public lastRemovedPermissions?: number | null
  public title?: string
  public department?: string
  public phoneNumber?: string

  constructor(person: IPerson) {
    this.altnetId = person.altnetId
    this.projectId = person.projectId
    this.userKind = person.userKind
    this.accessLevel = person.accessLevel
    this.name = person.name || {
      givenName: '',
      familyName: '',
      fullName: ''
    }
    this.primaryEmail = person.primaryEmail
    this.recoveryEmail = person.recoveryEmail
    this.phones = person.phones || []
    this.externalIds = person.externalIds || []
    this.orgUnitPath = person.orgUnitPath
    this.etag = person.etag
    this.isEnrolledInMFA = person.isEnrolledInMFA
    this.creationTime = person.creationTime
    this.lastLoginTime = person.lastLoginTime
    this.lastModifiedTime = person.lastModifiedTime
    this.notes = person.notes
    this.avatar =
      this.name.fullName === UI_STRINGS.RISKS.ANONYMOUS_USER ? { url: anonymousUserIcon, url_etag: '' } : person.avatar
    this.title = person.title
    this.department = person.department
    this.phoneNumber = person.phoneNumber
    this.internal = person.internal
    this.anonymous = !person.primaryEmail?.address || person.primaryEmail?.address?.indexOf('@') === -1 // altnetId: person.personId?.indexOf('@') === -1 &&
    this.internalCount = person.internalCount
    this.externalCount = person.externalCount
    this.emails = person.emails || []
    this.accessCount = person.accessCount
    this.riskCount = person.riskCount
    this.lastRemovedPermissions = person.lastRemovedPermissions
  }

  public get displayName(): string {
    if (this.name.givenName && this.name.familyName) {
      return `${this.name.givenName} ${this.name.familyName}`
    } else if (this.anonymous) {
      return UI_STRINGS.RISKS.ANONYMOUS_USER
    } else if (this.primaryEmail?.address) {
      const emailFirst = this.primaryEmail?.address.split('@')
      return emailFirst && emailFirst.length > 0 ? emailFirst[0] : this.primaryEmail?.address
    } else {
      return this.primaryEmail?.address ?? '' // altnetId:
    }
  }

  /*
   * For transform value from Map to Object in griddle-react custom components, e.g. SimplePersonCell, ActorCell
   * This is a workaround solution, once we refactor the table to NOT use griddle-react, this method shall be gone
   */
  static fromMap(personMap: { get: (x: keyof IPerson) => any }): Person {
    return new Person({
      primaryEmail: personMap.get('primaryEmail') && (Object.fromEntries(personMap.get('primaryEmail')) as Email),
      recoveryEmail: personMap.get('recoveryEmail') && (Object.fromEntries(personMap.get('recoveryEmail')) as Email),
      avatar: personMap.get('avatar') && (Object.fromEntries(personMap.get('avatar')) as Avatar),
      name: personMap.get('name') && (Object.fromEntries(personMap.get('name')) as Name),
      internal: personMap.get('internal'),
      internalCount: personMap.get('internalCount'),
      externalCount: personMap.get('externalCount'),
      emails:
        personMap.get('emails') &&
        (Array.from(personMap.get('emails')).map((email: any) => Object.fromEntries(email)) as Email[]),
      accessCount: personMap.get('accessCount'),
      riskCount: personMap.get('riskCount'),
      altnetId: personMap.get('altnetId'),
      projectId: personMap.get('projectId'),
      accessLevel: personMap.get('accessLevel'),
      userKind: personMap.get('userKind'),
      phones:
        personMap.get('phones') &&
        (Array.from(personMap.get('phones')).map((email: any) => Object.fromEntries(email)) as Phone[]),
      externalIds:
        personMap.get('externalIds') &&
        (Array.from(personMap.get('externalIds')).map((email: any) => Object.fromEntries(email)) as UserId[]),
      orgUnitPath: personMap.get('orgUnitPath'),
      etag: personMap.get('etag'),
      isEnrolledInMFA: personMap.get('isEnrolledInMFA'),
      creationTime: personMap.get('creationTime'),
      lastLoginTime: personMap.get('lastLoginTime'),
      lastModifiedTime: personMap.get('lastModifiedTime'),
      notes: personMap.get('notes') && (Object.fromEntries(personMap.get('notes')) as Notes)
    })
  }
}

export default Person
