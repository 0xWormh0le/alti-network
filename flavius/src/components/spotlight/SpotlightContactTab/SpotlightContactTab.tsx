import React from 'react'
import cx from 'classnames'
import Person from 'models/Person'
import Typography, { TypographyVariant } from 'components/elements/Typography'
import PersonAvatar from 'components/elements/PersonAvatar'
import CopyEmail from 'components/spotlight/CopyEmail'
import Marker, { MarkerType } from 'components/elements/Marker'
import { truncateText } from 'util/helpers'
import { UsageKind } from 'types/common'
import './SpotlightContactTab.scss'

export interface SpotlightContactTabProps {
  emailInfo: Email
  isInternal: boolean
  isSelected: boolean
  person: Person
}

const SpotlightContactTab: React.FC<SpotlightContactTabProps> = ({ emailInfo, isInternal, isSelected, person }) => {
  const { H4 } = TypographyVariant
  const { address } = emailInfo

  const personInfo =
    person.primaryEmail?.address === address
      ? new Person({ ...person })
      : new Person({
          ...person,
          name: {
            givenName: '',
            familyName: '',
            fullName: ''
          },
          primaryEmail: {
            address,
            kind: UsageKind.personal,
            primary: true,
            riskCount: 0,
            accessCount: 0,
            lastDeletedPermissions: 0
          }
        })

  return (
    <div
      key={address}
      className={cx('SpotlightContactTabItem', {
        'SpotlightContactTabItem--selected': isSelected,
        external: !isInternal
      })}>
      <PersonAvatar
        style={{ fontSize: '0.75em', marginRight: '1em' }}
        person={personInfo}
        className='SpotlightContactTabItem__avatar'
        clickable={false}
      />
      {isSelected ? (
        <CopyEmail email={address ?? ''}>
          <Typography variant={H4} component='div' className='SpotlightContactTabItem__email-text'>
            {truncateText(address ?? '', isInternal ? 22 : 18)}
          </Typography>
        </CopyEmail>
      ) : (
        <Typography variant={H4} component='div' className='SpotlightContactTabItem__email-text'>
          {truncateText(address ?? '', isInternal ? 22 : 18)}
        </Typography>
      )}
      {emailInfo.kind === UsageKind.personal && (
        <Marker type={MarkerType.EmailExteranl} className='SpotlightContactTabItem__external-marker' />
      )}
    </div>
  )
}

export default SpotlightContactTab
