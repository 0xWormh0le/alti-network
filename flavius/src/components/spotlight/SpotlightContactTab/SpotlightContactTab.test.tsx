import React from 'react'
import SpotlightContactTab, { SpotlightContactTabProps } from './SpotlightContactTab'
import { renderWithRouter } from 'test/support/helpers'
import Person from 'models/Person'
import { UsageKind, UserKind, AccessLevel } from 'types/common'

jest.mock('react', () => {
  const ActualReact = jest.requireActual('react')
  return {
    ...ActualReact,
    useContext: () => ({ domains: ['email.com'] })
  }
})

const person = new Person({
  name: {
    givenName: 'test',
    familyName: 'user',
    fullName: 'test user'
  },
  primaryEmail: {
    address: 'test@email.com',
    kind: UsageKind.personal,
    primary: true,
    accessCount: 0,
    riskCount: 0,
    lastDeletedPermissions: 0
  },
  recoveryEmail: {
    address: 'test@email.com',
    kind: UsageKind.personal,
    primary: true,
    accessCount: 0,
    riskCount: 0,
    lastDeletedPermissions: 0
  },
  avatar: {
    url: 'avatar',
    url_etag: 'avatar'
  },
  accessCount: 0,
  riskCount: 0,
  emails: [],
  internal: true,
  internalCount: 1,
  externalCount: 1,
  altnetId: '',
  projectId: '',
  accessLevel: AccessLevel.member,
  userKind: UserKind.user,
  phones: [],
  externalIds: [],
  orgUnitPath: '',
  etag: '',
  isEnrolledInMFA: false,
  creationTime: 0,
  lastLoginTime: 0,
  lastModifiedTime: 0,
  notes: {
    content: '',
    content_type: ''
  }
})

describe('SpotlightContactTab', () => {
  it('renders correctly', () => {
    const props: SpotlightContactTabProps = {
      person,
      emailInfo: person.primaryEmail,
      isInternal: false,
      isSelected: false
    }
    const { container } = renderWithRouter(<SpotlightContactTab {...props} />)
    expect(container).toMatchSnapshot()
  })
})
