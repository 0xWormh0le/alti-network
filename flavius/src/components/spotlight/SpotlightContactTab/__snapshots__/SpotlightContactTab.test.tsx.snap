// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotlightContactTab renders correctly 1`] = `
<div>
  <div
    class="SpotlightContactTabItem external"
  >
    <span
      class="PersonAvatar SpotlightContactTabItem__avatar trigger"
    >
      <div
        class="BaseAvatar "
        style="font-size: 0.75em; margin-right: 1em;"
      >
        <div
          class="BaseAvatar__image"
        >
          <img
            alt="tu"
            src="avatar"
            style="width: 2em;"
          />
        </div>
      </div>
    </span>
    <div
      class="Typography Typography--h4 SpotlightContactTabItem__email-text"
    >
      test@email.com
    </div>
    <span
      class="Typography Typography--body SpotlightContactTabItem__external-marker Marker Marker__email_external"
    >
      External
    </span>
  </div>
</div>
`;
